File: banquet-pilot/candidates/levels/enumerate_candidates.py
#!/usr/bin/env python3
"""Enumerate solutions for candidate JSON under candidates/levels/ only.

Copies the product solver API usage pattern without modifying product
levels/, dist/, or src/. Loads sibling c01.json–c06.json and prints counts.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

CAND_DIR = Path(__file__).resolve().parent
ROOT = CAND_DIR.parents[1]  # banquet-pilot/
sys.path.insert(0, str(ROOT / "src"))

from core.rules import validate_level  # noqa: E402
from core.solver import count_solutions, enumerate_solutions  # noqa: E402


def load_candidate(stem: str) -> dict:
    path = CAND_DIR / f"{stem}.json"
    with path.open(encoding="utf-8") as f:
        data = json.load(f)
    errors = validate_level(data)
    if errors:
        raise ValueError(f"invalid candidate {data.get('id', stem)}: {'; '.join(errors)}")
    return data


def report(stem: str, label: str | None = None, calm: set | None = None) -> tuple[int, int]:
    data = load_candidate(stem)
    tag = label or data.get("id", stem.upper())
    kwargs = {}
    if calm is not None:
        kwargs["calm"] = calm
    n, t = count_solutions(data, **kwargs)
    print(f"{tag}: {t} perms → {n} solution(s)")
    for s in enumerate_solutions(data, **kwargs):
        print(f"  {s}")
    return n, t


def main() -> int:
    print("=== Banquet Pilot candidate enumeration (candidates/levels only) ===")
    print(f"dir: {CAND_DIR}")

    report("c01")
    report("c02")

    c03 = load_candidate("c03")
    n0, t0 = count_solutions(c03, calm=set())
    print(f"C03 no calm: {t0} perms → {n0} solution(s)")
    n1, t1 = count_solutions(c03, calm={"rabbit"})
    print(f"C03 calm rabbit: {t1} perms → {n1} solution(s)")
    for s in enumerate_solutions(c03, calm={"rabbit"}):
        print(f"  {s}")
    for cid in ("fox", "crane", "otter", "tanuki", "hedgehog"):
        n, _ = count_solutions(c03, calm={cid})
        print(f"C03 calm {cid}: {n} solution(s)")

    report("c04")
    report("c05")
    report("c06")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: banquet-pilot/candidates/levels/enumerate_candidates.mjs
/** Enumerate candidate JSON under candidates/levels/ only (no product levels/). */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { validateLevel } from "../../src/core/rules.js";
import { countSolutions, enumerateSolutions } from "../../src/core/solver.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

function loadCandidate(stem) {
  const filePath = path.join(__dirname, `${stem}.json`);
  const data = JSON.parse(fs.readFileSync(filePath, "utf8"));
  const errors = validateLevel(data);
  if (errors.length > 0) {
    throw new Error(`invalid candidate ${data.id || stem}: ${errors.join("; ")}`);
  }
  return data;
}

function report(stem, label, calm) {
  const data = loadCandidate(stem);
  const tag = label || data.id || stem.toUpperCase();
  const opts = calm !== undefined ? { calm } : {};
  const [n, t] = countSolutions(data, opts);
  console.log(`${tag}: ${t} perms → ${n} solution(s)`);
  for (const s of enumerateSolutions(data, opts)) console.log(`  ${JSON.stringify(s)}`);
  return [n, t];
}

console.log("=== Banquet Pilot candidate enumeration (JS, candidates/levels only) ===");
console.log(`dir: ${__dirname}`);

report("c01");
report("c02");

{
  const c03 = loadCandidate("c03");
  const [n0, t0] = countSolutions(c03, { calm: new Set() });
  console.log(`C03 no calm: ${t0} perms → ${n0} solution(s)`);
  const [n1, t1] = countSolutions(c03, { calm: new Set(["rabbit"]) });
  console.log(`C03 calm rabbit: ${t1} perms → ${n1} solution(s)`);
  for (const s of enumerateSolutions(c03, { calm: new Set(["rabbit"]) })) {
    console.log(`  ${JSON.stringify(s)}`);
  }
  for (const cid of ["fox", "crane", "otter", "tanuki", "hedgehog"]) {
    const [n] = countSolutions(c03, { calm: new Set([cid]) });
    console.log(`C03 calm ${cid}: ${n} solution(s)`);
  }
}

report("c04");
report("c05");
report("c06");

{
  const c07 = loadCandidate("c07");
  const [n0, t0] = countSolutions(c07, { calm: new Set() });
  console.log(`C07 no calm: ${t0} perms → ${n0} solution(s)`);
  for (const cid of ["rabbit", "tanuki", "fox", "crane", "otter", "hedgehog"]) {
    const [n, t] = countSolutions(c07, { calm: new Set([cid]) });
    console.log(`C07 calm ${cid}: ${t} perms → ${n} solution(s)`);
    if (n > 0 && (cid === "rabbit" || cid === "tanuki")) {
      for (const s of enumerateSolutions(c07, { calm: new Set([cid]) })) {
        console.log(`  ${JSON.stringify(s)}`);
      }
    }
  }
}

report("c08");
report("c09");
